src/services/factories/getTrades.ts

```diff
@@ -52,44 +49,73 @@ export function getTradesFactory(factoryParams: {
     }
 
     const blocksSet = new Set<number>()
-    const orderIdsSet = new Set<string>()
+    const orderIdsMap = new Map<string, { buyTokenId: number; sellTokenId: number; blockNumber: number }>()
     const tokenIdsSet = new Set<number>()
 
     tradeEvents.forEach(event => {
-      blocksSet.add(event.blockNumber)
-      orderIdsSet.add(event.orderId)
-      tokenIdsSet.add(event.buyTokenId)
-      tokenIdsSet.add(event.sellTokenId)
+      const { blockNumber, sellTokenId, buyTokenId, orderId } = event
+
+      blocksSet.add(blockNumber)
+      orderIdsMap.set(orderId, {
+        buyTokenId,
+        sellTokenId,
+        blockNumber,
+      })
+      tokenIdsSet.add(buyTokenId)
+      tokenIdsSet.add(sellTokenId)
     })
 
-    // Filter orders that we might already have
-    const orders = new Map<string, Order>(
-      existingOrders
-        .filter(order => orderIdsSet.has(order.id))
-        .map(order => {
-          orderIdsSet.delete(order.id)
-          return [order.id, order]
-        }),
+    // ### ORDERS ###
+    const orders = new Map<string, Order>()
+    // Add all existing orders to map. No problem to have more than what we need since it'll be picked by id later.
+    existingOrders.forEach(order => orders.set(order.id, order))
+
+    // Fetch from contract the ones we don't have locally, and add it to the map
+    await Promise.all(
+      Array.from(orderIdsMap.keys()).map(async orderId => {
+        // We already have this order, ignore
+        if (orders.has(orderId)) {
+          return
+        }
+        // Fetch order from contract
+        let order = await exchangeApi.getOrder({ userAddress, networkId, orderId })
+        // Load additional info
+        const orderInfo = orderIdsMap.get(orderId)
+
+        // In case order was deleted from the contract, try to get it from OrderPlacement events instead
+        if (isOrderDeleted(order) && orderInfo) {
+          const { buyTokenId, sellTokenId, blockNumber } = orderInfo
+          try {
+            order = await exchangeApi.getOrderFromOrderPlacementEvent({
+              userAddress,
+              networkId,
+              orderId,
+              // Parameters bellow not required, but help narrow down the search
+              buyTokenId,
+              sellTokenId,
+              toBlock: blockNumber,
+            })
```

Comment:
There may be a better way
we can make calls to check contract state at a specific block
You have `blockNumber` from `TradeEvent` and `Order` corresponding to that event must be active at that point, of course. So you can do
```ts
contract.methods.getEncodedUserOrders(userAddress).call({}, blockNumber)
// then you'll have to decode orders as usual and find the one you need
```
But maybe even better since we have paginated orders and know `orderId`
```ts
exchangeApi.getOrdersPaginated({
 userAddress.
 networkId,
 offset: orderId,
 pageSize:1,
}, blockNumber) // fetches that exact order as it was at that block
```

I wonder if to have fewer fetches we could, if we have multiple deleted orders close by in id, fetch them all together with appropriate `pageSize`, but probably not, cause will be hard to guess at what `blockNumber` they all would be alive.

This is much cheaper (probably) than getting a bunch of unindexed events.

`blockNumber` isn't typed though, [it's not even in the docs](https://github.com/ethereum/web3.js/pull/3558)
So try with `//@ts-ignore` first and if it works for you, once again I'll need to hack around types in **dex-js**, ugh...

Reply:
Well, if that's the case, I could use the `getOrder` directly with the `blockNumber`.
Then orders will never be empty and I can skip this additional step entirely.

I'll give this a try :+1:

Reply:
Yes, that does work!

No need to check for this event then, I'll remove the method `getOrderFromOrderPlacementEvent` and this extra hop.

And, as expected, the build does fail:

```
ERROR in /home/alfeto/work/dfusion/dex-react/src/api/exchange/ExchangeApi.ts
ERROR in /home/alfeto/work/dfusion/dex-react/src/api/exchange/ExchangeApi.ts(347,83):
TS2554: Expected 0-1 arguments, but got 2.
```
![screenshot_2020-06-02_09-25-08](https://user-images.githubusercontent.com/43217/83544534-e5842c80-a4b2-11ea-9b81-6a7463c80a53.png)

Reply:
Several layers of ignore later :D
1cb96712bb95ae8063aeaeff7876978418f579f1

I'm wondering if I should still leave the check for empty orders there as an additional check.

But maybe not. There shouldn't be a case that a Trade event is emitted while an order does not exist on the contract.